src/btree/bt_delete.c

```diff
@@ -433,53 +641,44 @@ __wt_delete_page_instantiate(WT_SESSION_IMPL *session, WT_REF *ref, WT_PAGE_DELE
     WT_RET(__wt_page_modify_init(session, page));
     __wt_page_modify_set(session, page);
 
-    /* Allocate the per-page update array if one doesn't already exist. */
-    if (page->entries != 0 && page->modify->mod_row_update == NULL)
-        WT_PAGE_ALLOC_AND_SWAP(
-          session, page, page->modify->mod_row_update, upd_array, page->entries);
-
     /*
-     * Copy the page-deleted structure's timestamp information into an update for each row on the
-     * page. If the page-deleted structure is NULL, that means the truncate is globally visible, and
-     * therefore committed.
-     *
-     * If the truncate operation is not yet resolved, link updates in the page-deleted structure so
-     * they can be found when the transaction is resolved, even if they have moved to other pages.
+     * If the truncate operation is not yet resolved, count how many updates we're going to need and
+     * allocate an array for them. This allows linking them in the page-deleted structure so they
+     * can be found when the transaction is resolved, even if they have moved to other pages. If the
+     * page-deleted structure is NULL, that means the truncate is globally visible, and therefore
+     * committed. Use an extra slot to mark the end with NULL so we don't need to also store the
+     * length.
      */
     if (page_del != NULL && !page_del->committed) {
         count = 0;
-        WT_ROW_FOREACH (page, rip, i)
-            ++count;
+        switch (page->type) {
+        case WT_PAGE_COL_VAR:
+            /* One tombstone for each recno on the page. */
+            count = (uint32_t)(__col_var_last_recno(ref) - ref->ref_recno + 1);
```

Comment:
Just to check my own understanding of what's going on here - this (and the equivalent for row-store below) are what guarantees that `update_list[(*countp)++]` will always land in the array somewhere, right? And RLE doesn't need to be accounted for since we just allocate more than we need if some keys have been condensed?

Reply:
Yes to the first. Regarding the second: we need one tombstone per recno regardless of RLE compaction, since in-memory updates apply only to single keys. (That's why the update structure in VLCS is a whole insert list instead of just an update list; it needs to be keyed so that updates to different members of the same RLE cell can be distinguished.)

Reply:
OK, I think I understand now - thanks once again for the explanation.